### DISCUSSION TITLE
giving up :(

### DISCUSSION THREAD

**Discussion by sebastianp95** [vercel/next.js#52423] [App Router]
I had quite an unfavorable experience with Next.js 13, particularly when I attempted to integrate it with my Java server. While constructing static projects was fairly straightforward, it was a different story altogether as I ventured into server-side configurations.

The biggest obstacle I encountered was with cookie handling and Context API utilization in my root layout. Next.js cookie implementation proved to be unnervingly complex for a task that I initially perceived as simple. I found myself needing to create a 'route' component just to facilitate communication from my client component, which only contributed to the already jumbled setup.

Furthermore, when trying to incorporate Context into my root layout, it repeatedly triggered an endless loop regardless of my attempts to resolve it. It's a significant drawback in an otherwise capable framework. It's not a defeat per se, but I find it unfortunate that these are the challenges that got the best of me in Next.js 13. The experience underscores the need for improvements in the user-friendliness of some of its features.

**Comment by robertpiosik** [comment]
Could you prepare some minimal reproduction of your solution? "Context API utilization in root layout" is particularly alarming. Context API is a concept that lives on the client.

I strongly believe we can come up with a solution that will change your mind about the app router;-)

**Reply by sebastianp95** [reply]
Thank you for taking the time to consider my issue. Here are the two cases I've been struggling with, both of which result in an infinite loop of re-renders.

In the first example, I utilize DictionaryProvider:

```
export default async function RootLayout({ children, params }) {
  const dict = await getDictionary(params.lang);

  return (
    <html lang={params.lang}>
      <body className={inter.className}>
        <DictionaryProvider dict={dict}>
          <Navbar />
          {children}
        </DictionaryProvider>
      </body>
    </html>
  );
}
```

Here is the DictionaryProvider for reference:

```
"use client";
import React, { createContext } from "react";

export const DictionaryContext = createContext(null);

function DictionaryProvider({ children, dict }) {
  console.log("dict", dict);
  return (
    <DictionaryContext.Provider value={{ dict }}>
      {children}
    </DictionaryContext.Provider>
  );
}

export default DictionaryProvider;
```


And, I tried a different approach using AuthProvider:

```
export default async function RootLayout({ children, params }) {
  return (
    <html lang={params.lang}>
      <body className={inter.className}>
        <AuthProvider>
          <Navbar />
          {children}
        </AuthProvider>
      </body>
    </html>
  );
}
```

My ultimate aim was to harmoniously utilize both DictionaryProvider and AuthProvider together. If you could shed some light on this conundrum, I would appreciate it.

**Reply by robertpiosik** [reply]
Instead of the root layout, you can access `lang` param in a layout of a `[lang]` route.

**Reply by sebastianp95** [reply]
My root layout happens to be in the [lang] route. Do you suggest using the provider in a layout other than the root one? If so, would that mean I need to add a separate layout and provider for each page I create?
Even if I manage to do what you've suggested, I'll still need to wrap my app with my AuthProvider, which seems to be causing an infinite loop too. 
I could wrap all my pages in a folder 'containers' and here create a nested layout... is that what you mean?

Here's my project structure. Hope it helps

src\app\[lang]
src\app\[lang]\api
src\app\[lang]\dashboard
src\app\[lang]\homepage
src\app\[lang]\globals.css
src\app\[lang]\layout.js
src\app\[lang]\page.js

**Reply by robertpiosik** [reply]
What you're looking for are [route groups](https://nextjs.org/docs/app/building-your-application/routing/route-groups#convention). This way you can nest layouts, preserving a structure of your router.

**Comment by polarisdigitech** [comment]
Session management is a big pain in the ass in Next.js 13. Had similar issues like you, had to jump through hoops to do something very simple. And I don't even understand why server routes cannot live alongside view routes.

**Reply by robertpiosik** [reply]
If you explain what you were up to and how you tackled it we could try to help minimize the pain.

I don't know what exactly you have on your mind by server/view routes. When using static css like css modules you can absolutely have 'server routes' serving views, indeed.

**Reply by polarisdigitech** [reply]
For example, I have an application where authentication is handled by a different server that provides jwt. So if a user visits `/very-private-resource` which requires a trip to the server that handles authentication, and if the user's refresh token has expired, they need to login again.

Given cookies cannot be mutated in server components, so I can't do the following:

```ts
export default async function PrivateResource() {
    // Request to this external service might respond with a 403
    const data = await requestToExternalService();

    if (data == userTokenExpired) {
        cookies().delete('auth'); // Cannot perform this operation in a server component
        redirect('/login');
    }

...
```

So what I end up doing is either make an api call to a server route i.e `/api/invalidate-session` that invalidates the user's session, and then redirect to the login page after that request. Or append a query parameter i.e ` redirect('/login?invalidate')`, then use a middleware to check if the invalidate query parameter is present, then delete the user session.

Session can only be managed in server routes i.e `route.(js|ts)` or a middleware.

**Reply by robertpiosik** [reply]
Hmm, well, cookies cannot be mutated on the server in general so its not a good reason for complaining.
What about getting new access token every time user access `/login`? This way you don't need any logic with invalidation and login form would show up only when a refresh token is no longer valid.

**Reply by polarisdigitech** [reply]
> cookies cannot be mutated on the server in general

That isn't exactly correct, cookies can be mutated in server routes i.e `route.js`

```ts
const res = NextResponse.json({ message: "Login successful" });
res.cookies.set("auth", JSON.stringify(result));
```

I already know it can't be done on the server in Next.js hence why I'm complaining. You just basically said why I'm complaining isn't a good reason to complain without telling me why I shouldn't complain. That isn't helpful in anyway, given this is something can can be easily done with something like express, sveltekit etc. Not like I'm asking for something unreasonable.

In regards to your suggestion, the problem with that is the session is never invalidated. Which means when the user clicks the logout button they're not logged out. They can still access parts of my application even though they effectively can't perform any operation (because the external service will reject any operation). The application will look buggy to users and even internal stakeholders, cause a user shouldn't even see the `/private-resource` page to start with or any other page that should be behind the login.
